dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -754,58 +693,76 @@ void DMFileReader::addSkippedRows(UInt64 rows)
+        bool break_all_match = !pack_res[pack_id].allMatch() && read_rows >= rows_threshold_per_read / 2;
 
-            ++count;
-            rows += pack_stats[i].rows;
+        if (!is_use)
+        {
+            if (read_rows > 0)
+                read_block_infos.emplace_back(start_pack_id, pack_id - start_pack_id, last_pack_res, read_rows);
+            start_pack_id = pack_id + 1;
+            read_rows = 0;
+            last_pack_res = RSResult::All;
         }
-        return std::make_pair(count, rows);
-    };
-
-    for (size_t i = 0; i < pack_res.size();)
-    {
-        if (!pack_res[i].allMatch())
+        else if (reach_limit || break_all_match)
         {
-            ++i;
-            continue;
+            if (read_rows > 0)
+                read_block_infos.emplace_back(start_pack_id, pack_id - start_pack_id, last_pack_res, read_rows);
+            start_pack_id = pack_id;
+            read_rows = pack_stats[pack_id].rows;
+            last_pack_res = pack_res[pack_id];
+        }
+        else
+        {
+            last_pack_res = last_pack_res && pack_res[pack_id];
+            read_rows += pack_stats[pack_id].rows;
         }
-        auto [pack_count, rows] = get_all_match_block(i);
-        // Do not read block too small, it may hurts performance
-        if (rows >= rows_threshold_per_read / 2)
-            all_match_block_infos.emplace(i, pack_count);
-        i += pack_count;
     }
+    if (read_rows > 0)
+        read_block_infos.emplace_back(start_pack_id, pack_res.size() - start_pack_id, last_pack_res, read_rows);
 }
 
-size_t DMFileReader::getReadPackLimit(size_t start_pack_id)
+size_t DMFileReader::updateReadBlockInfos(const IColumn::Filter & filter, size_t pack_start, size_t pack_end)
 {
-    if (all_match_block_infos.empty() || read_one_pack_every_time)
-        return read_one_pack_every_time ? 1 : std::numeric_limits<size_t>::max();
-
-    const auto [next_all_match_block_start_pack_id, pack_count] = all_match_block_infos.front();
-    // Read packs with RSResult::All
-    if (next_all_match_block_start_pack_id == start_pack_id)
+    const auto & pack_stats = dmfile->getPackStats();
+    std::vector<std::tuple<size_t, size_t, RSResult, size_t>> new_read_block_infos;
+    new_read_block_infos.reserve(pack_end - pack_start);
+    size_t start_pack_id = pack_start;
+    const size_t start_row_offset = pack_id_to_offset[pack_start];
+    size_t read_rows = 0;
+    for (size_t pack_id = pack_start; pack_id < pack_end; ++pack_id)
     {
-        all_match_block_infos.pop();
-        return pack_count;
+        if (countBytesInFilter(filter, pack_id_to_offset[pack_id] - start_row_offset, pack_stats[pack_id].rows) == 0)
+        {
+            if (read_rows > 0)
+                // rs_result is not important here, we can use RSResult::All
+                new_read_block_infos.emplace_back(start_pack_id, pack_id - start_pack_id, RSResult::All, read_rows);
```

Comment:
> rs_result is not important here, we can use RSResult::All

Because `readWithFilter` does not consider the value of rs_result in the read_block_info? But I still think returning a RSResult::Some is better here.

Reply:
`readWithFilter` will use `rs_result` in the original read block info rather than the new one.